coderd/httpmw/provisionerdaemon.go

```diff
@@ -0,0 +1,69 @@
+package httpmw
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/codersdk"
+)
+
+type provisionerDaemonContextKey struct{}
+
+// ProvisionerDaemon returns the daemon from the ExtractProvisionerDaemon handler.
+func ProvisionerDaemon(r *http.Request) database.ProvisionerDaemon {
+	user, ok := r.Context().Value(provisionerDaemonContextKey{}).(database.ProvisionerDaemon)
+	if !ok {
+		panic("developer error: provisioner daemon middleware not provided")
+	}
+	return user
+}
+
+// ExtractWorkspaceAgent requires authentication using a valid provisioner token.
+func ExtractProvisionerDaemon(db database.Store) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			cookie, err := r.Cookie(codersdk.SessionTokenKey)
+			if err != nil {
+				httpapi.Write(rw, http.StatusUnauthorized, codersdk.Response{
+					Message: fmt.Sprintf("Cookie %q must be provided.", codersdk.SessionTokenKey),
+				})
+				return
+			}
+			token, err := uuid.Parse(cookie.Value)
+			if err != nil {
+				httpapi.Write(rw, http.StatusUnauthorized, codersdk.Response{
+					Message: "Provisioner token is invalid.",
+				})
+				return
+			}
+			provisioner, err := db.GetProvisionerDaemonByAuthToken(r.Context(), uuid.NullUUID{
+				Valid: true,
+				UUID:  token,
+			})
+			if err != nil {
+				if errors.Is(err, sql.ErrNoRows) {
+					httpapi.Write(rw, http.StatusUnauthorized, codersdk.Response{
+						Message: "Provisioner token is invalid.",
```

Comment:
This message is a bit misleading, token is a valid (format), but not registered or revoked. Simply saying `Forbidden.` could work.

Reply:
Like [my other comment](https://github.com/coder/coder/pull/3563#discussion_r958702580), this is basically just mimicking the way we handle agent tokens. I think if we get a token that isn't equal to the `auth_token` of a valid provisioner, we should generate the same error message regardless of whether it happens to be formatted like a UUID. 

But I'm totally open to changing the wording if there's a better way to describe that situation than "invalid".

Reply:
@spikecurtis I agree from a security perspective (don't reveal too much), but from a usability perspective I think it could be more helpful to the user. But I'm fine with either or.